### PR TITLE
Support converting certain <rubricBlock> elements to distractor rationale (WBL-63)

### PR DESCRIPTION
**Pull request by rakeshqdsa**
Some QTI assessment items will have content in <rubricBlock> elements. Some of those we map to shared passages, but some should be mapped to Learnosity metadata.distractor_rationale_response_level for students to see.

For more details Please see the following JIRA Ticket -
https://learnosity.atlassian.net/browse/WBL-63

### CODE REVIEW

**Comment by michaelsharman** [comment]
As per my last comment on the ticket, `distractor_rationale_response_level`should be inside the question, not the item. Eg:

```
{
    "options": [
        {
            "label": "[Choice A]",
            "value": "0"
        },
        {
            "label": "[Choice B]",
            "value": "1"
        },
        {
            "label": "[Choice C]",
            "value": "2"
        },
        {
            "label": "[Choice D]",
            "value": "3"
        }
    ],
    "stimulus": "<p>[This is the stem.]</p>",
    "type": "mcq",
    "validation": {
        "scoring_type": "exactMatch",
        "valid_response": {
            "score": 1,
            "value": []
        }
    },
    "metadata": {
        "distractor_rationale_response_level": [
            "<div id=\"SC-4857069\"><p><span class=\"style-ee989b5820727b1e90d5b0a63380e39c\">Distractor:</span> This expression represents the quotient of a number and 2 added to 4.</p></div>",
            "<div id=\"SC-4857071\"><p><span class=\"style-ee989b5820727b1e90d5b0a63380e39c\">Distractor:</span> This expression represents the quotient of a number and 4 added to 2.</p></div>",
            "<div id=\"SC-4857075\"><p><span class=\"style-ee989b5820727b1e90d5b0a63380e39c\">Correct:</span> This expression represents the quotient of the sum of a number and 4, divided by 2.</p></div>",
            "<div id=\"SC-4857080\"><p><span class=\"style-ee989b5820727b1e90d5b0a63380e39c\">Distractor:</span> This expression represents the quotient of the difference of a number and 4, divided by 2.</p></div>"
        ]
    }
}
```

**Comment by rakeshqdsa** [comment]
Done

File: src/Processors/QtiV2/In/DistractorRationaleResponseMapper.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
+<?php
+namespace LearnosityQti\Processors\QtiV2\In;
```
[the remark this answers is not in the document]
Comment (rakeshqdsa) [reply src/Processors/QtiV2/In/DistractorRationaleResponseMapper.php]: Done

File: src/Processors/QtiV2/In/DistractorRationaleResponseMapper.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+
+use DOMElement;
+use LearnosityQti\Utils\QtiMarshallerUtil;
```
[the remark this answers is not in the document]
Comment (rakeshqdsa) [reply src/Processors/QtiV2/In/DistractorRationaleResponseMapper.php]: Done

File: src/Processors/QtiV2/In/DistractorRationaleResponseMapper.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            }
+
+            $results['distractor_rationale_response_level'] = $distractorRationale;
+        } else {
+            throw new MappingException('No content found; cannot create distractor rational response level');
+        }
+        return $results;
+    }
+    
+    private function getDomForXml($xml)
```
[the remark this answers is not in the document]
Comment (rakeshqdsa) [reply src/Processors/QtiV2/In/DistractorRationaleResponseMapper.php]: Done
Create a new file in util and move function .

File: src/Processors/QtiV2/In/RubricBlockMapper.php [hunk 4]
```diff
@@ -66,6 +67,13 @@ public function parseWithRubricBlockComponent(RubricBlock $rubricBlock, $foundSc
                 }
                 break;
 
+            case ($rubricBlock->getUse() === 'rationale'):
```
[the remark this answers is not in the document]
Comment (rakeshqdsa) [reply src/Processors/QtiV2/In/RubricBlockMapper.php]: Done

File: src/Processors/QtiV2/In/DistractorRationaleResponseMapper.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        $isValid = $dom->loadXML($xml);
+
+        if (!$isValid) {
+            throw new MappingException('Invalid XML; Failed to parse DOM for sharedpassage content');
+        }
+
+        return $dom;
+    }
+    
+    private function getInnerXmlFragmentFromDom(\DOMDocument $dom)
```
[the remark this answers is not in the document]
Comment (rakeshqdsa) [reply src/Processors/QtiV2/In/DistractorRationaleResponseMapper.php]: Done

File: src/Processors/QtiV2/In/ItemBuilders/AbstractItemBuilder.php [hunk 6]
```diff
@@ -187,6 +190,11 @@ protected function setItemMetadata(array $itemMetadata)
         $this->metadata = array_merge_recursive($this->metadata, $itemMetadata);
     }
 
+    protected function setDistractorRationaleResponseLevel(array $distratorRationalResponseLevel)
```
[the remark this answers is not in the document]
Comment (rakeshqdsa) [reply src/Processors/QtiV2/In/ItemBuilders/AbstractItemBuilder.php]: Done

File: src/Processors/QtiV2/In/SharedPassageMapper.php [hunk 7]
```diff
@@ -1,20 +1,22 @@
 <?php
-
+include "../../../Utils/General/DOMHelper.php";
```
Comment (ttton) [inline src/Processors/QtiV2/In/SharedPassageMapper.php]: Why is this here? We should always be using the autoloader in here, we should never need to make includes. If this is related to the issue with the mismatched namespace and directory path on https://github.com/Learnosity/learnosity-qti/pull/55/files#diff-2cff45284c03e635f2e75e341a6c781eR3, I'd fix that instead.

File: src/Utils/General/DOMHelper.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
+<?php
+
+namespace LearnosityQti\Utils;
```
Comment (ttton) [inline src/Utils/General/DOMHelper.php]: This doesn't match the directory path of the class file, which includes the subspace `General`. You're going to have issues with the autoloader, as this is not PSR-4 compatible.

File: src/Utils/General/DOMHelper.php [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        // HACK: Pass the version and encoding to prevent libxml from decoding HTML entities (esp. &amp; which libxml borks at)
+        // Only do this if it hasnt already been passed along in the xml string. Sometimes, we read from a file, and sometimes
+        // we read from a block inside another file.
+        if (strpos($xml, '<?xml ') !== false) {
+            $xml = substr($xml, strpos($xml, '>') + 1);
+        }
+        $xml = '<?xml version="1.0" encoding="UTF-8">' . "<div>$xml</div>";
+
+        $dom->loadHTML($xml, LIBXML_HTML_NODEFDTD | LIBXML_HTML_NOIMPLIED);
+        $xml = $dom->saveXML($this->getFragmentWrapperDocumentElementForDom($dom));
```
Comment (ttton) [inline src/Utils/General/DOMHelper.php]: You haven't moved this `getFragmentWrapperDocumentElementForDom()` method into this class from https://github.com/Learnosity/learnosity-qti/pull/55/files#diff-b3f82e7247236692ccc3e7fa383883e2R238. This will break without it.

File: src/Utils/General/DOMHelper.php [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+        // HACK: Handle the fact that XML can't handle named entities (and HTML5 has no DTD for it)
+        $xml = XmlEntityUtil::convertNamedEntitiesToHexInString($xml);
```
Comment (ttton) [inline src/Utils/General/DOMHelper.php]: Needs a `use` for `XmlEntityUtil`

File: src/Processors/QtiV2/In/RubricBlockMapper.php [hunk 11]
```diff
@@ -72,20 +81,7 @@ public function parseWithRubricBlockComponent(RubricBlock $rubricBlock, $foundSc
...
-            case ($rubricBlock->getUse() === 'stimulus'):
-                if ($views->contains(View::CANDIDATE)) {
-                    $contents = QtiMarshallerUtil::marshallCollection($rubricBlock->getContent());
-                    $result = [
-                        'stimulus' => $contents,
-                    ];
-                }
-                break;
-
-            case ($rubricBlock->getClass() === 'DistractorRationale'):
```
Comment (ttton) [inline src/Processors/QtiV2/In/RubricBlockMapper.php]: We still need to support this key. It should be a fall-through case so we can handle both the same way, e.g.:

```
case ($rubricBlock->getUse() === 'rationale'):
    /* falls through */
case ($rubricBlock->getClass() === 'DistractorRationale'):
```

File: src/Processors/QtiV2/In/DistractorRationaleResponseMapper.php [hunk 12]
```diff
@@ -39,7 +37,7 @@ public function parseWithDistractorRationaleResponseComponent(RubricBlock $rubri
     
     private function getDomForXml($xml)
```
Comment (ttton) [inline src/Processors/QtiV2/In/DistractorRationaleResponseMapper.php]: It would make sense to me to also refactor this function so that we're using the `DOMHelper` class here too.

File: src/Processors/QtiV2/In/SharedPassageMapper.php (hunk 7 above)
Comment (rakeshqdsa) [reply src/Processors/QtiV2/In/SharedPassageMapper.php]: Done

File: src/Utils/General/DOMHelper.php (hunk 8 above)
Comment (rakeshqdsa) [reply src/Utils/General/DOMHelper.php]: Done

File: src/Utils/General/DOMHelper.php (hunk 9 above)
Comment (rakeshqdsa) [reply src/Utils/General/DOMHelper.php]: Done

File: src/Utils/General/DOMHelper.php (hunk 10 above)
Comment (rakeshqdsa) [reply src/Utils/General/DOMHelper.php]: Done

File: src/Processors/QtiV2/In/RubricBlockMapper.php (hunk 11 above)
Comment (rakeshqdsa) [reply src/Processors/QtiV2/In/RubricBlockMapper.php]: As per previous comment this is differentiated by views whether it is distractor rationale response level or only distractor rationale . 
Means use is rationale but views is candidate it would be for student and view is author than it would be for teacher .

File: src/Processors/QtiV2/In/DistractorRationaleResponseMapper.php (hunk 12 above)
Comment (rakeshqdsa) [reply src/Processors/QtiV2/In/DistractorRationaleResponseMapper.php]: Done

File: src/Processors/QtiV2/In/RubricBlockMapper.php (hunk 11 above)
Comment (ttton) [reply src/Processors/QtiV2/In/RubricBlockMapper.php]: @rakeshqdsa I'm not sure I follow. I know that we support two types of views separately for this scenario here. I'm talking about also handling both `use="rationale"` and `class="DistractorRationale"` cases. It is valid for QTI content to use either to achieve the same purpose. If we remove support for `class="DistractorRationale"`, then we break support for existing QTI content. Both of these cases need to behave the same way, hence my suggestion to include both cases.

Comment (rakeshqdsa) [reply src/Processors/QtiV2/In/RubricBlockMapper.php]: Done
